src/styles/cylc/_job.scss

```diff
@@ -0,0 +1,127 @@
+.job_theme--normal {
+
+    $teal: rgb(125,207,212);
+    $blue: rgb(106,164,241);
+    $green: rgb(81,175,81);
+    $red: rgb(207,72,72);
+    $pink: rgb(190,106,192);
+
+    .c-job {
+        svg.job {
+            rect {
+                /* if no job status display nothing */
+                fill: transparent;
+                stroke: transparent;
+            }
+
+            &.submitted rect {
+                fill: $teal;
+                stroke: $teal;
+            }
+
+            &.running rect {
+                fill: $blue;
+                stroke: $blue;
+            }
+
+            &.succeeded rect {
+                fill: $green;
+                stroke: $green;
+            }
+
+            &.failed rect {
+                fill: $red;
+                stroke: $red;
+            }
+
+            &.submit-failed rect {
+                fill: $pink;
+                stroke: $pink;
+            }
+        }
+    }
+}
+
+.job_theme--greyscale {
+
+    $light: rgb(208,208,208);
+    $medium: rgb(133,133,133);
+    $dark: rgb(0,0,0);
+
+    .c-job {
```

Comment:
With the approach of theming in css/scss, we will end up with a lot of repetition like the `.c-job` being re-defined as child of `.job_theme--themeN`, and may lead to maintenance problems if we end up customizing more than just colors.

Could we hold back theming, and implement in each component after https://github.com/cylc/cylc-ui/issues/9? Vuetify provides a way to have themes in [JS](https://github.com/cylc/cylc-ui/blob/aa6e75244dd4add7537552e4baf7fe9d904ebe27/src/plugins/theme.js), but very simple.

I think the best approach for `cylc-ui` would be to have the theme data in JS, but with Vuex. Similar to [this example](https://codesandbox.io/s/3no663nv6).

Reply:
The job theme isn't related to the app theme (users will need to be able to change the two separately) so this really is un-related to #9.

I'm not convinced by theming in JS but do try to convince me, I think in-line style is a much bigger maintenance headache then well structured CSS.

Reply:
>The job theme isn't related to the app theme (users will need to be able to change the two separately)

Agreed.

>I'm not convinced by theming in JS but do try to convince me, I think in-line style is a much bigger maintenance headache then well structured CSS.

When the user switches the themes, we will need to store the theme chosen in Vuex. Then I believe we would also use that value to dynamically set the value in `.job_theme--$value`.

So up to here both approaches would work and are quite similar.

1. As a user, 

I would like to be able to easily create themes and customize them in my site. With the CSS approach, they would need to edit SCSS and re-build the project.

With the JS approach, I **think** (would need to investigate) we could have dynamically loaded themes. Users woulds copy a file like `$cylc-ui-src/src/plugins/default-theme.js`, edit the values and put the JS somewhere in the server with a name `my-theme-name.js`, then the theme `my-theme-name` could be selected.

2. As a developer,

Either JS or CSS, I would find it easier if we had the definition of colours/variables/etc in one place, with the definition of elements/classes/etc somewhere else and without repetition.

Isn't there a way to achieve the same in CSS without the need to re-define classes under each theme? If the `.c-job` got more attributes or nested types, or if a component needs to define several classes, that will get annoyingly long.

If we are able to organize the current CSS code without this repetition, we can merge it and work around the CSS vs. JS approach in a separate issue. WDYT?

Reply:
Good points!

> Isn't there a way to achieve the same in CSS without the need to re-define classes under each theme

I think so, depends how scss compiles down (I'm used to sylus), will take a crack.

> Users woulds copy a file like `$cylc-ui-src/src/plugins/default-theme.js`

I guess the big advantage is that the user could create a custom theme from within the browser without having to work in the FS.

> we can merge it and work around the CSS vs. JS approach in a separate issue. WDYT?

I'm perfectly happy for this code to get scrapped either now or in the future, whatever's best.
(was only working on this because it's unlikely to get into conflict with other current work).

Reply:
>I'm perfectly happy for this code to get scrapped either now or in the future, whatever's best.
(was only working on this because it's unlikely to get into conflict with other current work).

So let's get it merged and discuss the theming elsewhere if you prefer. Besides the comment above :point_up: about `--fix` in the linter (which I think is not necessary), I think this is good to be merged.

Reply:
OK, I'm trying to work out how to to re-write the SCSS, I could put the job theme as a SCSS variable but then the CSS would have to be recompiled to change theme...

Reply:
Bit of hacking, the best I can come up with is:

```scss
/* normal theme */
$n-teal: rgb(125,207,212);
$n-blue: rgb(106,164,241);
$n-green: rgb(81,175,81);
$n-red: rgb(207,72,72);
$n-pink: rgb(190,106,192);

$job-theme-normal: (
    "submitted": (
        "fill": $n-teal,
        "stroke": $n-teal
    ),
    "running": (
        "fill": $n-blue,
        "stroke": $n-blue
    ),
    "suceeded": (
        "fill": $n-green,
        "stroke": $n-green
    ),
    "failed": (
        "fill": $n-red,
        "stroke": $n-red
    ),
    "submit-failed": (
        "fill": $n-pink,
        "stroke": $n-pink
    ),
);


/* greyscale theme */
$g-light: rgb(208,208,208);
$g-medium: rgb(133,133,133);
$g-dark: rgb(0,0,0);

$job-theme-greyscale: (
    "submitted": (
        "stroke": $g-light
    ),
    "running": (
        "fill": $g-light,
        "stroke": $g-light
    ),
    "succeeded": (
        "fill": $g-medium,
        "stroke": $g-medium
    ),
    "failed": (
        "fill": $g-dark,
        "stroke": $g-dark
    ),
    "submit-failed": (
        "stroke" $g-dark
    )
);


/* colour-blind theme */
$b-grey: rgb(152,152,152);
$b-success: rgb(108,218,255);
$b-error: rgb(146,0,0);

$job-theme-colour-blind: (
    "submitted": (
        "stroke": $b-grey
    ),
    "running": (
        "fill": $b-grey,
        "stroke": $b-grey
    ),
    "succeeded": (
        "fill": $b-success,
        "stroke": $b-success
    ),
    "failed": (
        "fill": $b-error,
        "stroke": $b-error
    ),
    "submit-failed": (
        "stroke": $b-error
    ),
);


/* job theme registry */
$job-themes: (
    "normal": $job-theme-normal,
    "greyscale": $job-theme-greyscale,
    "colour-blind": $job-theme-colour-blind
);


/* dump themes */
@each $name, $theme in $job-themes {
    .c-job {
        svg.job {
            rect {
                /* if no job status display nothing */
                fill: transparent;
                stroke: transparent;
            }
            @each $state, $style in $theme {
                &.#{$state} rect {
                    @each $key, $value in $style {
                        #{$key}: $value;
                    }
                }
            }
        }
    }
}
```

Reply:
Let me know which way you'd prefer.

Reply:
This one looks better. But we still have the CSS class redefined within each theme. I think we can have the CSS class defined in one place, and then the theme variables defined elsewhere... e.g.

```scss
// we can define the colours of the default job-theme in a
// ~styles/cylc/job-themes/_colours.scss or _variables.scss...
// I used this one just to avoid creating global variables in this file,
// and because the colours are quite similar to the ones you picked...
@import "../material-dashboard/colors";

/* job css class wrapper */
.c-job {
    svg.job {
        rect {
            /* if no job status display nothing */
            fill: var(--job_theme--fill);
            stroke: var(--job_theme--stroke);
        }

        &.submitted rect {
            fill: var(--job_theme--submitted-fill);
            stroke: var(--job_theme--submitted-stroke);
        }

        &.running rect {
            fill: var(--job_theme--running-fill);
            stroke: var(--job_theme--running-stroke);
        }

        &.succeeded rect {
            fill: var(--job_theme--succeeded-fill);
            stroke: var(--job_theme--succeeded-stroke);
        }

        &.failed rect {
            fill: var(--job_theme--failed-fill);
            stroke: var(--job_theme--failed-stroke);
        }

        &.submit-failed rect {
            fill: var(--job_theme--submit-failed-fill);
            stroke: var(--job_theme--submit-failed-stroke);
        }
    }
}

// --- Theming with CSS variables [1], could be moved to theme file(s)?

/**
 The normal theme uses the same variables from the original theme.

 The other classes simply re-define these variables locally, in their own
 scope. Once a component adopts the class, the variables are replaced in
 "runtime". This way SCSS variables are used in "compile time" (dev), and
 CSS variables in "runtime" (live), with the option of being changed with
 CSS and/or JS.
 */

// instead of the putting variables in :root, another alternative is to
// provide a second argument to var($variable, $fallback), moving the
// default values to the fallback. Maybe that's better?

/* job_theme global variables */
:root {
    --job_theme--fill: transparent;
    --job_theme--stroke: transparent;
    --job_theme--submitted-fill: #{$teal}; // see note [2]
    --job_theme--submitted-stroke: #{$teal};
    --job_theme--running-fill: #{$blue};
    --job_theme--running-stroke: #{$blue};
    --job_theme--succeeded-fill: #{$green};
    --job_theme--succeeded-stroke: #{$green};
    --job_theme--failed-fill: #{$red};
    --job_theme--failed-stroke: #{$red};
    --job_theme--submit-failed-fill: #{$pink};
    --job_theme--submit-failed-stroke: #{$pink};
}

/* normal theme variable values */
.job_theme--normal {
    --job_theme--fill: transparent;
    --job_theme--stroke: transparent;
    --job_theme--submitted-fill: #{$teal};
    --job_theme--submitted-stroke: #{$teal};
    --job_theme--running-fill: #{$blue};
    --job_theme--running-stroke: #{$blue};
    --job_theme--succeeded-fill: #{$green};
    --job_theme--succeeded-stroke: #{$green};
    --job_theme--failed-fill: #{$red};
    --job_theme--failed-stroke: #{$red};
    --job_theme--submit-failed-fill: #{$pink};
    --job_theme--submit-failed-stroke: #{$pink};
}

/* greyscale theme variable values */
.job_theme--greyscale {
    $light: rgb(208,208,208);
    $medium: rgb(133,133,133);
    $dark: rgb(0,0,0);
    --job_theme--fill: transparent;
    --job_theme--stroke: transparent;
    --job_theme--submitted-fill: transparent;
    --job_theme--submitted-stroke: #{$light};
    --job_theme--running-fill: #{$light};
    --job_theme--running-stroke: #{$light};
    --job_theme--succeeded-fill: #{$medium};
    --job_theme--succeeded-stroke: #{$medium};
    --job_theme--failed-fill: #{$dark};
    --job_theme--failed-stroke: #{$dark};
    --job_theme--submit-failed-fill: transparent;
    --job_theme--submit-failed-stroke: #{$dark};
}

/* colour blind theme variable values */
.job_theme--colour_blind {
    $grey: rgb(152,152,152);
    $success: rgb(108,218,255);
    $error: rgb(146,0,0);
    --job_theme--fill: transparent;
    --job_theme--stroke: transparent;
    --job_theme--submitted-fill: transparent;
    --job_theme--submitted-stroke: #{$grey};
    --job_theme--running-fill: #{$grey};
    --job_theme--running-stroke: #{$grey};
    --job_theme--succeeded-fill: #{$success};
    --job_theme--succeeded-stroke: #{$success};
    --job_theme--failed-fill: #{$error};
    --job_theme--failed-stroke: #{$error};
    --job_theme--submit-failed-fill: transparent;
    --job_theme--submit-failed-stroke: #{$error};
}

// [1] https://www.bolajiayodeji.com/introducing-css-custom-properties-variables/
// [2] See issue https://github.com/sass/sass/issues/1128, need string interpolation
//     here instead of accessing SCSS variable directly

```

Reply:
This way I think users should be able to override the values in their own themes with some CSS... and maybe a JS somewhere that activates/loads the themes (still need to investigate this part, though we may be able to use CSS variables, which seem to be a compromise between our approaches?).

Not sure if we need to define the CSS [`!default`](https://stackoverflow.com/questions/10643107/what-does-default-in-a-css-property-value-mean) marker as well, to avoid users having to use `!important`... but still needs investigation as well.

Reply:
Ooh, pure CSS3 variables. Will have a go.

Reply:
Issues with this approach are:

* Themes don't have freedom over the CSS (e.g. can't add a new property)
* The `:root` theme can leak into the job themes (imagine what will happen to custom themes if we add a new property)

Latest idea:

Define the css namespace `c-job svg.job rect` as a variable at the top of the file, use it to namespace theme CSS that way we aren't re-defining the inheritance only the fact that `.job` must live within `.job_theme` which is a fundamental assumption anyway.

Reply:
>Themes don't have freedom over the CSS (e.g. can't add a new property)

I thought that was expected. I mean, we would define the properties to be re-defined by users.

>The :root theme can leak into the job themes (imagine what will happen to custom themes if we add a new property)

Yup, see the TODO in the example above. Instead of pollyting root, we can simply use a default value in the `var(--..., $default-value-instead-of-root)`.

>As we are re-defining the variables, how does theme switching work?

A user would choose a new theme. That theme would be set in Vuex via JavaScript. The components would rely on Vuex to retrieve the application theme name, and concatenate the component base class (e.g. `job_theme--`) with the theme name.

>Define the css namespace c-job svg.job rect as a variable at the top of the file, use it to namespace theme CSS that way we aren't re-defining the inheritance only the fact that .job must live within .job_theme which is a fundamental assumption anyway.

Uh, that sounds interesting. Can you elaborate with a short code example? Would it be something like:

```scss
$theme-namespace: .c-job svg.job react;
```

And then `#{$theme-namespace}` in some parts of the CSS?

Reply:
> Uh, that sounds interesting. Can you elaborate with a short code example?

Commited as https://github.com/cylc/cylc-ui/pull/134/commits/3dfda938c9a03cf43920a6ab552371ddc0fd7b33

Reply:
Ah, similar to what I thought you suggested.

It reduces duplication in the code, but the final produced CSS is still the same. If we had to add a new state (e.g. `dead`). We would have to change the contents of `#{$cjob} {` under N themes (i.e. .job_theme--normal, .job_theme--greyscale, job_theme--...).

Whereas with the CSS variables we would change it in one place I think?

Reply:
> If we had to add a new state (e.g. dead). We would have to change the contents of #{$cjob} { under N themes

Yes! To add a new state we would need to modify each theme, that is the correct behaviour to me, we wouldn't want to fallback on a global definition.

Reply:
I think we have an impasse here. Perhaps a third person could come up with another solution, or convince one of us :) (but I feel like it's clarified how we both were thinking around theming, and we've made some good progress in the discussion)

Reply:
I guess this still needs to be resolved?

Reply:
Yup, no consensus on which way to implement theming. But I think I digressed too much, so perhaps we should use @oliver-sanders ' approach and move the discussion around how to do theming for another issue.

@oliver-sanders feel free to update the PR with your latest version (if you still want to modify it). I can test it again, and then approve it if there are no errors on console/IDE/etc :+1:

Reply:
I'm happy to wait for your theming PR then adapt this into it.

Reply:
Ah, wait, you've closed #143....